### PR TITLE
Fix Stripe Billing checkout with different currencies

### LINKED ISSUE #7732: Nondescript error with Stripe Billing if customers attempt to have multiple active subscriptions with different currencies
### Describe the bug
If a site uses Stripe Billing with WooPayments + Woo Subscriptions and a customer has an active subscription in one currency (e.g., EUR), they cannot purchase another subscription in a different currency (e.g., USD).

This is expected, but it produces a nondescript error on checkout:

```
There was a problem creating your subscription. Please try again or contact us for assistance.
```

Additionally, there are no subscription notes for the `Pending` subscription or the `Draft` Parent Order that are created. As a result, no one will know what happened unless:

1. Logging is enabled.
2. A merchant assesses the logs

The logs indicate the following:

```
You cannot combine currencies on a single customer. This customer has an active subscription, subscription schedule, discount, quote, or invoice item with currency eur. (invalid_request_error)
```

### To Reproduce

1. Enable Stripe Billing.
2. Enable logging.
3. Enable multi-currency.
4. Add at least one additional currency.
5. Access the store while logged into an account (e.g., your admin account).
6. Complete a purchase for a subscription product in one currency.
7. Reproduce step 5 while logged into the same account as step 4, but change the currency.
8. Note the `There was a problem` error.
9. Check the order and subscription notes.
10. Check the logs.

### Actual behavior
The error provided to the customer has zero indication as to why the payment failed, and there is no information surfaced to the merchant unless logging is enabled _and_ they access the logs.

### Screenshots
blocks invalid request error

order notes

sub notes

### Expected behavior
The `invalid_request_error` is surfaced to either the customer or the merchant in a way that is actionable.

### PR DESCRIPTION
**Pull request by reykjalin**
Fixes #7732

## Changes proposed in this Pull Request

* Create new `Cannot_Combine_Currencies_Exception`.
* Throw this exception when the following conditions are met:
    1. The merchant has Stripe Billing enabled;
    2. The merchant has multi-currency enabled with at least 2 available currencies;
    3. The customer has a pre-existing, active subscription with currency A;
    4. The customer tries to buy a new subscription with currency B; and
    5. The Stripe API returns an error indicating that a customer attempted to checkout with a new currency.
* Catch the exception in the subscription service and throw an error that informs the customer of why the payment failed.

### Setup

1. Checkout `develop`.
2. Make sure you have WC Subscriptions installed on your test store.
3. Make sure you have a subscription product available, a simple subscription is sufficient.
4. Enable Stripe Billing in the WooPayments settings.
5. Enable multi-currency and enable any 2 currencies (e.g. USD and EUR)

### Reproduce

1. Add a subscription product to your cart, and checkout with any card, `4242 4242 4242 4242` is sufficient.
2. The checkout should succeed.
3. Change the currency.
    * You can do this either by using the currency switching widget or by visiting any store page with `?currency=<currency_code>`, e.g. `?currency=EUR`.
4. Add the same subscription product to your cart and checkout with any card, e.g. the card that was saved after checking out in (1).
5. Read the error message:

| | Before (`develop`) | After |
|:---:|:---:|:---:|
| Shortcode | image | image |
| Blocks | image | image |

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-7.5.0#make-sure-an-actionable-error-is-reported-when-trying-to-buy-a-subscription-through-stripe-billing-with-a-different-currency-from-your-active-subscriptions
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable. **Not applicable**
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8462` or branch name `fix/stripe-billing-checkout-with-different-currencies` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/stripe-billing-checkout-with-different-currencies) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 053cb5de7a231604ae05ba04f0e0c76056e70700
- Build time: 2024-04-03 16:46:07 UTC

_Note: the build is updated when a new commit is pushed to this PR._

**Comment by reykjalin** [comment]
@ricardo When I run the tests individually - either via CLI filter or by using `it.only` - they run to completion but when I run them together they always fail. Running them with `npm run test:e2e-dev` causes a segfault in Chromium.

I have no idea why. Can you see if you can run the e2e tests properly?

**Comment by reykjalin** [comment]
Marked with `needs design` because I'd like to get design to review the language in the error message. Asked in p1711762260578259-slack-C01TNBRTMEU.

File: includes/subscriptions/class-wc-payments-subscription-service.php [hunk 1]
```diff
@@ -417,6 +418,15 @@ public function create_subscription( WC_Subscription $subscription ) {
 						'</strong>'
 					)
 				);
+			} elseif ( $e instanceof Cannot_Combine_Currencies_Exception ) {
+				throw new Exception(
+					sprintf(
+						// Translators: %1$s and %2$s are both currency codes, e.g. `USD` or `EUR`.
+						__('There was a problem creating your subscription. All your active subscriptions must use the same currency. You attempted to purchase a subscription in %1$s but have another active subscription using %2$s.', 'woocommerce-payments' ),
```
Comment (ricardo) [inline includes/subscriptions/class-wc-payments-subscription-service.php]: 🤏 Nit: Missing space after opening parenthesis.

**Comment by ricardo** [comment]
> Can you see if you can run the e2e tests properly?

@reykjalin The new test spec looks good, but I couldn't get my test environment to work with Stripe billing.

When I run `npm run test:e2e-dev tests/e2e/specs/subscriptions/shopper/shopper-subscriptions-multi-currency.spec.js`, it fails with the following error for both tests:

> There was a problem creating your subscription. Please try again or contact us for assistance.

This is probably due to some issue with my local e2e environment, because the PR tests well on my Valet WooPayments environment.

**Review by ricardo — APPROVED** [review]
Tests well and LGTM.

I get the following error notice when trying to pay with a different currency:

Stripe Billing checkout with different currencies Notice

The only thing to mention is that I also couldn't run the e2e tests as you requested. Maybe we could ask someone else from Fractal before merging? I need more time to troubleshoot why my local e2e tests env is not working.

File: includes/subscriptions/class-wc-payments-subscription-service.php (hunk 1 above)
Comment (reykjalin) [reply includes/subscriptions/class-wc-payments-subscription-service.php]: Hmmm, I'm realizing PHPCS doesn't catch this anymore 🤔 

Not even if I revert 660a6302dc439aed8751c547952d8724647bb475 to go back to before we updated the sniffs.

Not sure why. Huh.

Comment (reykjalin) [reply includes/subscriptions/class-wc-payments-subscription-service.php]: Fixed in bcd7fdc3a2957726fb46cafe883bd2aec180e3dc

Comment (reykjalin) [reply includes/subscriptions/class-wc-payments-subscription-service.php]: Opened #8533 to track the phpcs issue

**Review by ricardo — APPROVED** [review]
LGTM 👍.

**Pull request merged**